### PR TITLE
expose a way to add custom http logging filters

### PR DESCRIPTION
**Pull request by spennihana**
- for easier testing, use Mockito (new dependency) to mock HttpServlet params

### CODE REVIEW

File: h2o-core/build.gradle [hunk 1]
```diff
@@ -42,6 +42,7 @@ dependencies {
   // compile 'net.java.dev.jets3t:jets3t:0.6.1'
 
   testCompile "junit:junit:${junitVersion}"
+  testCompile "org.mockito:mockito-core:1.+"
```
Comment (mmalohlava) [inline h2o-core/build.gradle]: Minor: it is better to define an exact version number.

Comment (spennihana) [inline h2o-core/build.gradle]: thanks, i spec'd 2.6.2

File: h2o-core/src/main/java/water/api/RequestServer.java [hunk 2]
```diff
@@ -80,6 +80,16 @@
   public static ArrayList<Route> routes() { return routesList; }
   public static Route lookupRoute(RequestUri uri) { return routesTree.lookup(uri, null); }
 
+  private static HttpLogFilter[] _filters=null;
+  public static void setFilters(HttpLogFilter... filters) {
+    _filters=filters;
+  }
+
+  static {
```
Comment (tomkraljevic) [inline h2o-core/src/main/java/water/api/RequestServer.java]: -1 for static blocks

Comment (spennihana) [inline h2o-core/src/main/java/water/api/RequestServer.java]: I am not sure what the source of the objection is (there could be several), but perhaps one solution is a bit that is flipped that prevents filters from being set after registration is finalized (similar to the _acceptRequests bit used by jetty)

If it is some other issue, may you please specify?

**Comment by tomkraljevic** [comment]
no, my only objection is the static block...
we want fewer static blocks (trending to 0, so unit testability will be possible someday), not more...

the idea is fine.

tom

**Comment by spennihana** [comment]
Ok cool, got rid of the static block

**Comment by spennihana** [comment]
Hi,

Is there more that I can do to move this change through?

Thanks, Spencer

**Comment by mmalohlava** [comment]
@h2o-ops please test!

**Comment by mmalohlava** [comment]
Sorry for delay Spencer, I just triggered tests for changes, if they pass (relevant tests) i am going to merge it.

**Comment by spennihana** [comment]
Hi Michal,
Any insight into the failing checks?

Thanks!
Spencer

**Comment by navdeep-G** [comment]
@spennihana @mmalohlava Looks like the test links went stale :(. I will rerun.

**Comment by navdeep-G** [comment]
@h2o-ops please test!

**Comment by spennihana** [comment]
Hi, it does not appear that the tests were rerun?

Clicking the "details" from this page attempts to send me to what looks like an internal IP address:
e.g. http://172.17.2.151:8080/job/PR_mojo_compatibility/38/

Thanks,
Spencer

**Comment by mmalohlava** [comment]
@h2o-ops please test

**Comment by spennihana** [comment]
For the h2o_auto_runit_small job:
I see this failure:


12:48:15 Conducting R package/version check...
12:48:18 Error in pkgs_check_update(pkgs, check_only = check_only, repos = repos) : 
12:48:18   Check unsuccessful.
12:48:18 Execution halted
12:48:18 
12:48:18 ERROR: /home/ubuntu/workspace/h2o_auto_runit_small/h2o-r/scripts/package_version_check_update.R failed.
12:48:18        See /home/ubuntu/workspace/h2o_auto_runit_small/h2o-r/tests/results/package_version_check_out.txt

Additionally I see this:

12:48:28 Total tests:               532
12:48:28 Passed:                    0
12:48:28 Did not pass:              0
12:48:28 Did not complete:          532
12:48:28 H2O INTERNAL tests:        1
12:48:28 H2O INTERNAL failures:     1
12:48:28 NOPASS tests (not run):    35
12:48:28 NOFEATURE tests (not run): 9



So my test is failing some internal test?

I am unable to examine the other jobs.

**Comment by spennihana** [comment]
OK, sorry for the spam on this thread!

I am performing ./gradlew syncRPackages as per failing check -- perhaps this will get my fork up-to-date and pass remaining checks, I will push once it's thru syncing packages

**Comment by spennihana** [comment]
orthogonal issue: Why does R have SO many dependencies now? is this just for testing? this sync is taking a VERY long time:

INFO: Jenkins (package,version) list:
(data.table, 1.10.1)
(acepack, 1.3-3.3)
(ade4, 1.7-2)
(AUC, 0.3.0)
(BH, 1.58.0-1)
(bit, 1.1-12)
(bit64, 0.9-5)
(bitops, 1.0-6)
(boot, 1.3-17)
(brew, 1.0-6)
(caTools, 1.17.1)
(class, 7.3-14)
(cluster, 2.0.3)
(codetools, 0.2-14)
(colorspace, 1.2-6)
(crayon, 1.3.1)
(curl, 0.9.3)
(DEoptimR, 1.0-3)
(devtools, 1.9.1)
(dichromat, 2.0-0)
(digest, 0.6.8)
(diptest, 0.75-7)
(e1071, 1.6-7)
(evaluate, 0.8)
(flexclust, 1.3-4)
(flexmix, 2.3-13)
(foreach, 1.4.2)
(foreign, 0.8-66)
(Formula, 1.2-1)
(fpc, 2.1-10)
(gbm, 2.1.1)
(gdata, 2.17.0)
(ggplot2, 1.0.1)
(git2r, 0.11.0)
(glmnet, 2.0-2)
(gplots, 2.17.0)
(gridExtra, 2.0.0)
(gtable, 0.1.2)
(gtools, 3.5.0)
(HDtweedie, 1.1)
(Hmisc, 3.17-0)
(httr, 1.0.0)
(iterators, 1.0.7)
(jsonlite, 0.9.17)
(kernlab, 0.9-22)
(KernSmooth, 2.23-15)
(labeling, 0.3)
(lattice, 0.20-33)
(latticeExtra, 0.6-26)
(LiblineaR, 1.94-2)
(magrittr, 1.5)
(MASS, 7.3-45)
(Matrix, 1.2-3)
(mclust, 5.0.2)
(memoise, 0.2.1)
(mgcv, 1.8-10)
(mime, 0.4)
(mlbench, 2.1-1)
(modeltools, 0.2-21)
(munsell, 0.4.2)
(mvtnorm, 1.0-3)
(nlme, 3.1-124)
(nnet, 7.3-11)
(plyr, 1.8.3)
(prabclus, 2.2-6)
(proto, 0.3-10)
(R6, 2.1.1)
(randomForest, 4.6-10)
(RColorBrewer, 1.1-2)
(Rcpp, 0.12.2)
(RCurl, 1.95-4.7)
(reshape2, 1.4.1)
(rgl, 0.95.1441)
(R.methodsS3, 1.7.0)
(robustbase, 0.92-5)
(ROCR, 1.0-7)
(R.oo, 1.19.0)
(roxygen2, 4.1.1)
(rpart, 4.1-10)
(rstudioapi, 0.3.1)
(RUnit, 0.4.29)
(R.utils, 2.1.0)
(rversions, 1.0.2)
(scales, 0.3.0)
(spatial, 7.3-11)
(statmod, 1.4.21)
(stringi, 1.0-1)
(stringr, 1.0.0)
(survival, 2.38-3)
(testthat, 0.10.0)
(trimcluster, 0.1-2)
(whisker, 0.3-2)
(xml2, 0.1.2)

INFO: Missing the following Jenkins-approved R packages:
spatial
proto
randomForest
boot
rgl
ade4
RUnit
AUC
mlbench
HDtweedie
LiblineaR
statmod
bit
bit64
R.methodsS3
R.oo
R.utils
bitops
RCurl
caTools
KernSmooth
gtools
gdata
gplots
ROCR
codetools
iterators
foreach
lattice
Matrix
nlme
mgcv
glmnet
modeltools
flexclust
MASS
nnet
flexmix
DEoptimR
robustbase
trimcluster
mclust
kernlab
diptest
mvtnorm
prabclus
cluster
class
e1071
fpc
foreign
acepack
Formula
rpart
colorspace
munsell
dichromat
labeling
mime
R6
rstudioapi
git2r
brew
whisker
magrittr
stringi
BH
RColorBrewer
gtable
Rcpp
digest
xml2
curl
rversions
jsonlite
gridExtra
survival
gbm
latticeExtra
stringr
evaluate
roxygen2
memoise
crayon
testthat
httr
devtools
plyr
reshape2
scales
ggplot2
Hmisc
data.table


this seems like a jenkins/build issue, and not something i'm missing from my fork? i.e., these checks seem broken.

**Comment by navdeep-G** [comment]
@h2o-ops please test!

**Comment by mmalohlava** [comment]
I am not sure about so huge number of packages - from @jangorecki i heard there were some changes in r-packages we depend on which introduced a number of new dependencies. But it is worth to explore it.

**Comment by mmalohlava** [comment]
@h2o-ops please test!

**Comment by jangorecki** [comment]
@spennihana yes, those packages are being used during h2o-3 build and tests. h2o R package deps are already stripped down to minimum and are listed in package DESCRIPTION file.

**Pull request merged**